src/toplevel.pl

```diff
@@ -326,62 +282,163 @@
     write('"w": write terms without depth limit\n'),
     write('"p": print terms with depth limit\n\n').
 
-gather_query_vars([_ = Var | Vars], QueryVars) :-
-    (  var(Var) ->
-       QueryVars = [Var | QueryVars0],
-       gather_query_vars(Vars, QueryVars0)
-    ;
-       gather_query_vars(Vars, QueryVars)
-    ).
-gather_query_vars([], []).
+terms_equations(Eqs0, AllInitVs, InterestVs0, AnonVs0, ResVs0, NewVs0, Terms, AllEqs) :-
+    % Include new variables of interest, possibly some anonymous variables.
+    term_variables(InterestVs0, InterestVs),
+
+    % Include new anonymous variables. New variables of anonymous origin are
+    % new anonymous variables.
+    term_variables(AnonVs0, AnonVs1),
+
+    % Anonymous variables that are "inside" a variable of interest isn't
+    % anonymous.
+    % TODO: Make sure no new variable is added (`bagof/3` issue).
+    (   bagof(
+            V,
+            V^(
+                lists:member(V, AnonVs1),
+                \+ '$toplevel':eq_member(V, InterestVs)
+            ),
+            AnonVs) ->
+        true
+    ;   AnonVs = []
+    ),
 
-is_a_different_variable([_ = Binding | Pairs], Value) :-
-    (  Value == Binding, !
-    ;  is_a_different_variable(Pairs, Value)
-    ).
+    % Get the attributed variables only.
+    '$term_attributed_variables'(ResVs0, ResVs1), % Not enough.
+    /* The predicate `bagof/3` cannot be since it introduces new variables.
+    (   bagof(
+            V,
+            V^(lists:member(V, ResVs0), \+ '$toplevel':eq_member(V, AnonVs)),
+            ResVs1
+        ) ->
+        true
+    ;   ResVs1 = []
+    ),
+    % */
+    not_in_list(AnonVs, ResVs1, ResVs2),
+
+    term_variables([AllInitVs, ResVs1], AllVs),
+
+    /* The predicate `bagof/3` cannot be since it introduces new variables.
+    (   bagof(
+            V,
+            V^(lists:member(V, AllVs), \+ '$toplevel':eq_member(V, AnonVs)),
+            AttrVs
+        ) ->
+        true
+    ;   AttrVs = []
+    ),
+    % */
+    not_in_list(AnonVs, AllVs, AttrVs0),
+
+    % '$term_attributed_variables'(AttrVs0, AttrVs1),
+    % '$project_atts':project_attributes(AttrVs0, AttrVs0),
```

Comment:
This line should be uncommented and changed to
```
'$project_atts':project_attributes(AllInitVs, AttrVs0),
```

Reply:
But `AllInitVs` doesn't contain only variables. In a query like `?- Vs = [X, _], Vs in 0..3.`, before the query is executed `AllInitVs = [Vs, X, _A]`, after the execution `AllInitVs = [[X, _A], X, _A]`. What would be the effect?

Reply:
In that case, `AllInitVs` should be replaced with a list of its variables that remain unbound after the query is executed.

Reply:
Should `'$project_atts':project_attributes/2` be called on all the variables (stored in `AllVs` which include residue variables) or only variables in `AllInitVs`?

Reply:
It should be called on unbound query variables, so those of `AllInitVs`.

Reply:
Done but I'm still wondering about the attributed variable from `call_residue_vars/3` in `AttrVs0`.

Reply:
The example
```
?- Vs0 = [A, B, C], freeze(A, B = C), call_residue_vars(bagof(V, V^(member(V, Vs0)), Vs), Ws), copy_term(Vs, Vs, AttrGs).
    Vs0 = [A, B, C], Ws = [A], Vs = [A, B, C], AttrGs = [freeze:freeze(A, user:(B = C), user:(B = C), user:(B = C), user:(B = C))].
```
can be reduced to
```
?- Vs0 = [A, B, C], freeze(A, B = C), bagof(V, V^(member(V, Vs0)), Vs), Ws).
   Vs0=[A,B,C], Ws=[A,B,C], freeze:freeze(A,user:(B=C),user:(B=C),user:(B=C),user:(B=C)).
```
and the result is the same, after pruning a few variables from query.

When `A` is united to `B` and `C`, the frozen goal `B=C` is propagated to each of them. That's entirely due to the definition of `verify_attributes` in the `freeze` module.

Reply:
But with query like `Vs0 = [A, B, C], dif(A, B), bagof(V, V^(member(V, Vs1)), Vs).` or `Vs0 = [A, B, C], freeze(A, false), bagof(V, V^(member(V, Vs1)), Vs).` runs out of memory.

Reply:
`Vs1` is unbound in either case, so `member(V, Vs1)` succeeds endlessly.

Reply:
My bad. It's specific to `freeze` the issue. One way to fix `freeze` could be an extra test `Fa \= Fb`.

But it also happens with `copy_term/2`:
<pre>
<b>?- Vs = [A, B, C], freeze(A, B = C), copy_term(Vs, Vs).</b>
    Vs=[A,B,C], freeze:freeze(A,user:(B=C),user:(B=C)). % Unexpected.
<b>?- Vs = [A, B, C], freeze(A, B = C), copy_term(Vs, Vs, Gs).</b>
    Vs=[A,B,C], Gs=[freeze:freeze(A,user:(B=C))].
<b>?- Vs0 = [A, B, C], freeze(A, B = C), copy_term(A, X).</b>
    Vs0=[A,B,C], freeze:freeze(A,user:(B=C)), freeze:freeze(X,user:(_A=_B)).
<b>?- Vs0 = [A, B, C], freeze(A, B = C), copy_term(A, X), A = X.</b>
    Vs0=[A,B,C], X=A, freeze:freeze(A,user:(_A=_B)). % Unexpected by definition of `freeze`.
<b>?- </b>
</pre>
But attributed variables are being trigger while I didn't expect it.

Reply:
`copy_term/2` and `copy_term/3` work a bit differently. `copy_term/3` copies its first argument to the second argument without copying the first's attributes. 

`copy_term/2` copies attributes. The act of unifying `Vs` with its copy triggers `freeze/2`'s `verify_attributes`, which redundantly copies the freeze goal to `A`'s attributes.

In your second unexpected example, unifying `X` to `A` doesn't call `B = C` because `X` and `A` are both variables.

Reply:
I can reconcile with `bagof/3` yet. The first query was right, the last query I can not reproduce it.
```
?- Vs0 = [A, B, C], freeze(A, B = C), copy_term(A, X), A = X.
    Vs0=[A,B,C], X=A, freeze:freeze(A,user:(B=C),user:(_A=_B)).
?-
```